File: midterm/sorting.cpp
#include <iostream>
#include <vector>
#include <algorithm>
# include <cmath>


using namespace std;

// Bubble Sort
// 정렬되지 않은 값 중 최댓값을 맨 뒤로 보내기 반복
// 인접한 두 값을 반복하여 비교 및 교환
void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
} 




// Selection Sort
// 정렬되지 않은 값들 중에서 최솟값 찾아 앞에 놓기 반복
// Unstable 함 !!
void selectionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        swap(arr[i], arr[minIdx]);
    }
}





// Insertion Sort
// 정렬되지 않은 값들 중 하나를 골라서,  이미 정렬된 값들 사이에 끼워넣기 반복

// while문, 마지막에 값을 올바른 위치에 넣는 코드
void insertionSort(vector<int>& arr) {
    int n = arr.size();  // 배열의 크기를 n에 저장
    for (int i = 1; i < n; i++) {  // 두 번째 원소부터 배열 끝까지 반복
        int tmp = arr[i];          // 현재 정렬할 원소를 임시 변수 tmp에 저장
        int j = i - 1;          // 정렬된 부분의 마지막 원소 인덱스

        // tmp보다 큰 원소들을 한 칸씩 뒤로 이동
        while (j >= 0 && arr[j] > tmp) {
            arr[j + 1] = arr[j];  // 현재 원소를 뒤로 한 칸 이동
            j--;                // 인덱스 j를 하나씩 감소하여 이전 원소와 비교
        }
        arr[j + 1] = tmp;  // tmp 값을 올바른 위치에 삽입
    }
}


// 추천 !
// for문 2개, 바로바로 swap하는 버전 코드
void insertionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 1; i < n; i++) {  // 두 번째 원소부터 시작
        for (int j = i; j > 0; j--) {
            if (arr[j - 1] <= arr[j]) {
                break;  // 앞의 원소가 현재 원소보다 작거나 같으면 정렬 완료
            }
            swap(arr[j - 1], arr[j]);
        }  // 현재 원소가 더 작으면 두 원소를 교체
    }
}





// Merge Sort
// 분할 정복(Divide and Conquer) 정렬 알고리즘  
// Key Idea: 영역을 나눠서 각각 정렬하고 합친다  
// 나뉜 영역을 재귀적으로 정렬
// 배열의 두 부분을 병합하는 함수
void merge(vector<int>& arr, int l, int m, int r) {
    // cout << "merge"<<endl;
    // cout << "l "<<l <<",m "<<m<<" ,r "<<r<<endl;
    // 임시 배열 tmp에 arr의 부분 배열 [l, r) 복사
    vector<int> tmp(arr.begin() + l, arr.begin() + r);
    
    int i = 0, j = m - l, k = l;  // i: 왼쪽 부분, j: 오른쪽 부분, k: 병합할 arr의 인덱스
    
    // 왼쪽과 오른쪽 부분 배열을 비교하여 작은 값부터 arr에 다시 병합
    while (i < m - l && j < r - l) {
        // cout << i<< ":tmp[i] "<< tmp[i]<<" " <<j<<":tmp[j] " << tmp[j]<<endl;
        // cout << "k "<<k <<endl;
        if (tmp[i] < tmp[j]) {
            arr[k] = tmp[i];  // 왼쪽 값이 작으면 arr에 복사
            i += 1;
        } else {
            arr[k] = tmp[j];  // 오른쪽 값이 작거나 같으면 arr에 복사
            j += 1;
        }
        k += 1;  // 인덱스 k 증가
    }
    
    // 왼쪽 부분에 남은 값들을 arr에 복사
    while (i < m - l) {
        // cout<< "왼쪽 남는 ";
        // cout << i<< ":tmp[i] "<< tmp[i] << endl;
        arr[k] = tmp[i];
        i += 1;
        k += 1;
    }
    
    // 오른쪽 부분에 남은 값들을 arr에 복사
    while (j < r - l) {
        // cout<< "오른쪽 남는 ";
        // cout << j<< ":tmp[j] "<< tmp[j] << endl;
        arr[k] = tmp[j];
        j += 1;
        k += 1;
    }
}

// 배열을 분할하여 정렬하는 함수 (Merge Sort)
void _merge_sort(vector<int>& arr, int l, int r) {
    if (l + 1 == r) return;  // 부분 배열이 1개 원소면 더 이상 분할할 필요 없음
    
    int m = (l + r) / 2;  // 중간 인덱스 계산
    // cout << "mergesort "<<endl;
    // cout << "l "<<l <<",m "<<m<<" ,r "<<r<<endl;
    _merge_sort(arr, l, m);  // 왼쪽 절반 재귀적으로 정렬
    _merge_sort(arr, m, r);  // 오른쪽 절반 재귀적으로 정렬
    
    // 정렬된 두 부분을 병합
    merge(arr, l, m, r);
}


void mergeSort(vector<int>& arr) {
    _merge_sort(arr, 0, arr.size());
}






// Quick Sort
// 분할 정복(Divide and Conquer) 정렬 알고리즘  
// Key Idea: Pivot을 기준으로 영역을 나눠서 각각 정렬  
// 나뉜 영역을 재귀적으로 정렬 partition
// 배열을 분할하여 피벗을 기준으로 작은 값과 큰 값으로 나누는 함수

// // while문 사용 : 비추
int partition(vector<int>& arr, int left, int right) {
    int pivot = arr[(left + right) / 2];  // 피벗은 중간값으로 설정
    while (left <= right) {  // 좌측과 우측이 교차할 때까지 반복
        // 피벗보다 작은 값을 찾기 위해 좌측 인덱스를 증가
        while (arr[left] < pivot) left++;
        // 피벗보다 큰 값을 찾기 위해 우측 인덱스를 감소
        while (arr[right] > pivot) right--;
        // left와 right가 교차하지 않은 상태라면 두 값을 교환
        if (left <= right) {
            swap(arr[left], arr[right]);  // 좌측 값과 우측 값을 교환
            left++;  // 좌측 인덱스 증가
            right--;  // 우측 인덱스 감소
        }
    }
    return left;  // 교차된 이후의 좌측 인덱스를 반환
}

// Ver2. for문 사용 : 이걸로 공부!!!!
int partition(vector<int>& arr, int left, int right) {
    int pivot = arr[right];  // 피벗을 배열의 가장 오른쪽 값으로 설정
    int i = left;  // 작은 값들이 위치할 인덱스
    // cout <<"pivot "<<pivot <<", i "<< i<< endl;

    // left부터 right-1까지 배열을 순회
    for (int j = left; j < right; j++) {  // right를 제외한 범위
        // 현재 값이 피벗보다 작거나 같으면 i 위치에 교환
        if (arr[j] <= pivot) {
            // cout << i<< " : swap "<< arr[i]<<endl;
            // cout << j<< " : swap "<< arr[j]<<endl;
            swap(arr[i], arr[j]);
            i++;  // i를 증가시켜 작은 값의 위치를 이동
        }
    }

    // 마지막으로 피벗을 올바른 위치에 놓음
    swap(arr[i], arr[right]);

    return i;  // 피벗이 위치한 인덱스를 반환
}

// 퀵소트를 수행하는 함수
void quickSort(vector<int>& arr, int left, int right) {
    // cout << "quicksort" <<left << right <<endl;
    if (left < right) {
        int part = partition(arr, left, right);  // 배열을 분할하고 분할 위치를 얻음

        quickSort(arr, left, part - 1);  // 좌측 부분 재귀적으로 정렬
        quickSort(arr, part + 1, right);  // 우측 부분 재귀적으로 정렬
    }
}

// Ver3. 
//피벗 중앙값으로 설정
int partition(vector<int>& arr, int left, int right) {
    int mid = (left + right) / 2;  // 중간값 인덱스를 계산
    int pivot = arr[mid];  // 중간값을 피벗으로 설정

    // 피벗을 배열의 끝으로 이동 (right로 이동시켜 나중에 올바른 위치로 옮김)
    swap(arr[mid], arr[right]);

    int i = left;  // 작은 값들이 위치할 인덱스

    // left부터 right-1까지 배열을 순회
    for (int j = left; j < right; j++) {  // 피벗을 제외한 범위 탐색
        if (arr[j] <= pivot) {  // 피벗보다 작은 값들은 왼쪽으로
            swap(arr[i], arr[j]);
            i++;  // i를 증가시켜 작은 값의 위치를 이동
        }
    }

    // 마지막으로 피벗을 올바른 위치에 놓음
    swap(arr[i], arr[right]);  // 피벗을 제자리에 위치시킴

    return i;  // 피벗이 위치한 인덱스를 반환
}







// Heap Sort
void heapify(vector<int>& arr, int parent, int last) {
    int left, right, largest;
    while (parent * 2 + 1 <= last) {
        left = parent * 2 + 1;
        right = parent * 2 + 2;
        largest = parent;
        if (left <= last && arr[left] > arr[largest]) {
            largest = left;
        }
        if (right <= last && arr[right] > arr[largest]) {
            largest = right;
        }
        if (largest != parent) {
            swap(arr[parent], arr[largest]);
            parent = largest;
        } else {
            break;
        }
    }
}

void heapSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, i, n - 1);
    }
    for (int i = n - 1; i >= 0; i--) {
        swap(arr[0], arr[i]);
        heapify(arr, 0, i - 1);
    }
}



// Radix Sort
// 값들이 모두 k자릿수 이하의 음이 아닌 정수일때 r진수 (10진수)
// 낮은 자릿수LSD부터 높은 자릿수MSD 순서대로 자릿수를 기준으로 안정 정렬
void stable_sort_by_digit(vector<int>& arr, int k, int r) {
    int n = arr.size();
    vector<int> cnts(r, 0);
    int rtok = pow(r, k);

    for (int x : arr) {
        int digit = (x / rtok) % r;
        cnts[digit]++;
    }

    for (int i = 1; i < r; i++) {
        cnts[i] += cnts[i - 1];
    }

    vector<int> sorted_arr(n);
    for (int i = n - 1; i >= 0; i--) {
        int digit = (arr[i] / rtok) % r;
        sorted_arr[--cnts[digit]] = arr[i];
    }

    arr = sorted_arr;
}

void radixSort(vector<int>& arr, int k, int r) {
    for (int i = 0; i < k; i++) {
        stable_sort_by_digit(arr, i, r);
    }
}




// Counting Sort
// 값들이 r이하의 음이 아닌 정수일 때
// 각 값의 개수를 세어, 값마다 들어갈 위치를 미리 계산
// 계수 정렬(counting sort) 알고리즘의 개념

// 계수정렬은 숫자들간 비교를 하지 않고 정렬을 하는 알고리즘
// 일일이 비교를 하지 않고 각 숫자가 몇개인지 센 다음에 정렬함
// 배열내 최대 값+1 만큼의 길이만큼의 메모리가 필요함
void countingSort(vector<int>& arr, int maxValue) {
    int n = arr.size();
    vector<int> count(maxValue + 1, 0);
    vector<int> output(n);

    // 각 값의 출현 빈도를 카운트
    for (int i = 0; i < n; i++) {
        count[arr[i]]++;
    }

    // count 배열을 수정하여 각 값이 최종적으로 위치할 인덱스를 계산
    for (int i = 1; i <= maxValue; i++) {
        count[i] += count[i - 1];
    }

    // output 배열에 정렬된 결과 저장
    for (int i = n - 1; i >= 0; i--) {
        output[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }

    // 원래 배열에 정렬된 값을 복사
    for (int i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}




// main

int main() {
    //vector<int> arr = {101, 193, 237, 492, 395, 159};
    vector<int> arr = {4, 3, 9, 8, 7, 2, 5, 6};

    int mx = *max_element(arr.begin(), arr.end());

    //bubbleSort(arr);
    //selectionSort(arr);
    insertionSort(arr);
    //mergeSort(arr);
    //quickSort(arr, 0, arr.size()-1);
    // heapSort(arr);
    //radixSort(arr, 3, 10);
    // countingSort(arr, mx);

    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}

File: midterm/sorting.py
import random



# # bubble sort  Θ(n2)
# # 정렬되지 않은 값 중 최댓값을 맨 뒤로 보내기 반복
# # 인접한 두 값을 반복하여 비교 및 교환
# def bubble_sort(arr):
#     n = len(arr)
#     for a in range(n-1, 0, -1):
#         for i in range(a):
#             if arr[i] > arr[i+1] :
#                 arr[i], arr[i+1] = arr[i+1], arr[i]

 
# # selection sort Θ(n2)
# # 정렬되지 않은 값들 중에서 최솟값 찾아 앞에 놓기 반복
# def selection_sort(arr):
#     n = len(arr)
#     for i in range(n-1) : 
#         min_idx = i
#         min_n = arr[i]
#         for j in range(i+1, n):
#             if arr[j] < min_n :
#                 min_idx = j
#                 min_n = arr[j]
        
#         if min_idx != i :
#             arr[i], arr[min_idx] = arr[min_idx], arr[i]



# # insertion sort Θ(n2)
# # 정렬되지 않은 값들 중 하나를 골라서,  
# # 이미 정렬된 값들 사이에 넣고 뒤는 미루기 반복
# def insertion_sort(arr):
#     n = len(arr)
#     for j in range(n-1):
#         for i in range(n-1):
#             if arr[i] <= arr[i+1]: # 같거나 arr[i]가 더 작으면
#                 break
#             arr[i], arr[i+1] = arr[i+1], arr[i] # 뒤로 미루기




# # merge sort
# # 분할 정복(Divide and Conquer) 정렬 알고리즘  
# # Key Idea: 영역을 나눠서 각각 정렬하고 합친다  
# # 나뉜 영역을 재귀적으로 정렬

# def merge_sort(arr):
#     _merge_sort(arr, 0, len(arr))

# def _merge_sort(arr, l, r):
#     if l + 1 == r: return
#     m = (l + r) // 2
#     _merge_sort(arr, l, m)
#     _merge_sort(arr, m, r)
#     _merge(arr, l, m, r)

# def _merge(arr, l, m, r):
#     tmp = arr[l:r]
#     i, j, k = 0, m-l, l
#     while i < (m-l) and j < (r-l):
#         if tmp[i] < tmp[j]:
#             arr[k] = tmp[i]
#             i += 1; k += 1
#         else:
#             arr[k] = tmp[j]
#             j += 1; k += 1
#     while i < m-l:
#         arr[k] = tmp[i]
#         i += 1; k += 1
#     while j < r-l:
#         arr[k] = tmp[j]
#         j += 1; k += 1

# def merge_sort(arr):
#     # 배열 arr 전체를 정렬하는 함수
#     _merge_sort(arr, 0, len(arr))  # 재귀적 분할 정복 시작

# def _merge_sort(arr, l, r):
#     # 배열의 l부터 r까지 부분을 정렬하는 함수
#     if l + 1 == r: 
#         return  # 배열의 크기가 1이면 정렬할 필요가 없으므로 종료

#     m = (l + r) // 2  # 배열의 중간 인덱스 계산
#     _merge_sort(arr, l, m)  # 왼쪽 절반을 재귀적으로 정렬
#     _merge_sort(arr, m, r)  # 오른쪽 절반을 재귀적으로 정렬
#     _merge(arr, l, m, r)  # 정렬된 두 부분 배열을 병합

# ### 주석 있는 버전
# def _merge(arr, l, m, r):
#     # arr[l:m]과 arr[m:r] 두 정렬된 배열을 병합하는 함수
#     tmp = arr[l:r]  # arr[l:r]의 부분 배열을 복사해 tmp 배열에 저장
#     i, j, k = 0, m - l, l  # i는 왼쪽 배열, j는 오른쪽 배열의 시작점, k는 원래 배열에서의 현재 위치

#     # 두 배열의 값을 비교하면서 더 작은 값을 원래 배열 arr에 채워 넣음
#     while i < (m - l) and j < (r - l):
#         if tmp[i] < tmp[j]:  # 왼쪽 배열의 값이 더 작으면
#             arr[k] = tmp[i]  # 왼쪽 값을 arr에 저장
#             i += 1  # 왼쪽 배열에서 다음 값으로 이동
#         else:  # 오른쪽 배열의 값이 더 작거나 같으면
#             arr[k] = tmp[j]  # 오른쪽 값을 arr에 저장
#             j += 1  # 오른쪽 배열에서 다음 값으로 이동
#         k += 1  # arr에서 다음 위치로 이동

#     # 왼쪽 배열에 남아있는 값들을 arr에 채워 넣음
#     while i < m - l:
#         arr[k] = tmp[i]
#         i += 1
#         k += 1

#     # 오른쪽 배열에 남아있는 값들을 arr에 채워 넣음
#     while j < r - l:
#         arr[k] = tmp[j]
#         j += 1
#         k += 1





# # quick sort
# # 분할 정복(Divide and Conquer) 정렬 알고리즘  
# # Key Idea: Pivot을 기준으로 영역을 나눠서 각각 정렬  
# # 나뉜 영역을 재귀적으로 정렬 partition

def quick_sort(arr):
    quicksort(arr, 0, len(arr) - 1)

def quicksort(arr, l, r):
    if l < r:
        # 배열을 분할한 후 피벗 인덱스를 기준으로 재귀 호출
        p_index = partition(arr, l, r)
        quicksort(arr, l, p_index - 1)  # 왼쪽 부분 정렬
        quicksort(arr, p_index + 1, r)  # 오른쪽 부분 정렬

def partition(arr, l, r):
    # 피벗을 랜덤하게 선택
    pivot_index = random.randint(l, r)
    arr[pivot_index], arr[r] = arr[r], arr[pivot_index]  # 피벗을 마지막으로 이동

    pivot = arr[r]  # 피벗을 마지막 요소로 설정
    i = l - 1  # 피벗보다 작은 값들을 추적하는 인덱스
    print(pivot ,i)
    for j in range(l, r):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]  # 작은 값과 교체
    # 피벗을 작은 값과 큰 값 사이에 위치시킴

    arr[i + 1], arr[r] = arr[r], arr[i + 1]
    return i + 1  # 새로운 피벗 위치 반환








# heap sort

def heap_sort(arr):
    n = len(arr)  # 배열의 길이를 저장

    # 최대 힙을 구성하는 과정
    # 부모 노드는 자식 노드의 인덱스보다 항상 작아야 하므로,
    # 마지막 부모 노드부터 시작하여 힙을 구축
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)  # 현재 인덱스에서 최대 힙을 구성

    # 하나씩 꺼내서 최대 힙을 정렬하는 과정
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # 루트(최대 값)와 마지막 요소를 교환
        heapify(arr, i, 0)  # 교환된 부분을 제외한 나머지 힙을 재구성

def heapify(arr, n, i):
    largest = i  # 현재 노드의 인덱스를 저장
    l = 2 * i + 1  # 왼쪽 자식의 인덱스 계산
    r = 2 * i + 2  # 오른쪽 자식의 인덱스 계산

    # 왼쪽 자식이 현재 노드보다 크면 largest를 왼쪽 자식으로 업데이트
    if l < n and arr[l] > arr[largest]:
        largest = l
    
    # 오른쪽 자식이 현재 largest 값보다 크면 largest를 오른쪽 자식으로 업데이트
    if r < n and arr[r] > arr[largest]:
        largest = r

    # 만약 largest가 현재 노드가 아니라면
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # 현재 노드와 largest를 교환
        heapify(arr, n, largest)  # 교환된 노드에서 다시 힙 구조를 유지




# counting sort
# 값들이 r이하의 음이 아닌 정수일 때
# 각 값의 개수를 세어, 값마다 들어갈 위치를 미리 계산

def counting_sort(arr, r):
    n = len(arr)
    cnts = [0] * (r + 1)

    # 각 요소의 빈도를 계산
    for x in arr:
        cnts[x] += 1

    # 누적합 계산 (인덱스 1부터 시작)
    for i in range(1, r + 1):
        cnts[i] += cnts[i - 1]

    # 정렬된 배열을 만들기 위한 배열
    sorted_arr = [0] * n

    # 뒤에서부터 처리하여 정렬된 배열 채우기
    for x in reversed(arr):
        cnts[x] -= 1
        sorted_arr[cnts[x]] = x

    # 원본 배열에 정렬된 값 복사
    arr[:] = sorted_arr



# radix sort
# 값들이 모두 k자릿수 이하의 음이 아닌 정수일때 r진수 (10진수)
# 낮은 자릿수LSD부터 높은 자릿수MSD 순서대로 자릿수를 기준으로 안정 정렬
def radix_sort(arr, k, r):
    for i in range(k):
        stable_sort_by_digit(arr, i, r)

def stable_sort_by_digit(arr, k, r):
    n = len(arr)
    cnts = [0] * r
    rtok = r ** k  # k번째 자리수를 추출하기 위한 값

    for x in arr:
        digit = (x // rtok) % r  # k번째 자리수 추출
        cnts[digit] += 1

    for i in range(1, r):
        cnts[i] += cnts[i - 1]

    sorted_arr = [0] * n

    for x in reversed(arr):
        digit = (x // rtok) % r  # 다시 자리수 추출
        cnts[digit] -= 1
        sorted_arr[cnts[digit]] = x

    arr[:] = sorted_arr


arr = [101,193,237,492,395,159]
# #arr = [1,3,7,2,5,9]
# counting_sort(arr,500)
# #radix_sort(arr, 3, 10)
quick_sort(arr)
print(arr)







